File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ws=we=0
        ans=0
        D={}
        while(we<len(s)):
            if D.get(s[we],-1)==-1:
                D[s[we]]=we
                
                ans=max(ans,we-ws+1)
            else:
                temp=D[s[we]]+1
                while(ws<temp):
                    del D[s[ws]]
                    ws+=1     
                D[s[we]]=we
                
            we+=1
        return ans
            
                
                
                
        

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.rb
def length_of_longest_substring(s)
    n = s.length
    set = Set.new
    max_length, i, j = 0, 0, 0
    while i < n && j < n
        if !set.include?(s[j])
            set.add(s[j])
            j += 1
            max_length = [max_length, j - i].max
        else
            set.delete(s[i])
            i += 1
        end
    end
    return max_length
end